src/control/control.cpp

```diff
@@ -112,53 +112,69 @@ QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(
+                VERIFY_OR_DEBUG_ASSERT(!pCreatorCO) {
+                    qWarning()
+                            << "ControlObject"
+                            << key.group << key.item
+                            << "already created";
+                    return nullptr;
+                }
+                return pControl;
             } else {
-                pControl = it.value();
+                // The weak pointer has become invalid and can be cleaned up
+                s_qCOHash.erase(it);
             }
         }
     }
 
-    if (pControl == NULL) {
-        if (pCreatorCO) {
-            pControl = QSharedPointer<ControlDoublePrivate>(
-                    new ControlDoublePrivate(key, pCreatorCO, bIgnoreNops,
-                                             bTrack, bPersist, defaultValue));
-            MMutexLocker locker(&s_qCOHashMutex);
-            //qDebug() << "ControlDoublePrivate::s_qCOHash.insert(" << key.group << "," << key.item << ")";
-            s_qCOHash.insert(key, pControl);
-        } else if (!flags.testFlag(ControlFlag::NoWarnIfMissing)) {
-            qWarning() << "ControlDoublePrivate::getControl returning NULL for ("
-                       << key.group << "," << key.item << ")";
-            DEBUG_ASSERT(flags.testFlag(ControlFlag::NoAssertIfMissing));
-        }
+    if (pCreatorCO) {
+        auto pControl = QSharedPointer<ControlDoublePrivate>(
+                new ControlDoublePrivate(key,
+                        pCreatorCO,
+                        bIgnoreNops,
+                        bTrack,
+                        bPersist,
+                        defaultValue));
+        const MMutexLocker locker(&s_qCOHashMutex);
+        //qDebug() << "ControlDoublePrivate::s_qCOHash.insert(" << key.group << "," << key.item << ")";
+        s_qCOHash.insert(key, pControl);
+        return pControl;
+    }
+
+    if (!flags.testFlag(ControlFlag::NoWarnIfMissing)) {
+        qWarning() << "ControlDoublePrivate::getControl returning NULL for ("
+                   << key.group << "," << key.item << ")";
+        DEBUG_ASSERT(flags.testFlag(ControlFlag::NoAssertIfMissing));
     }
-    return pControl;
+    return nullptr;
 }
 
 // static
 void ControlDoublePrivate::getControls(
-        QList<QSharedPointer<ControlDoublePrivate> >* pControlList) {
-    s_qCOHashMutex.lock();
+        QList<QSharedPointer<ControlDoublePrivate> >* pControlList,
+        bool detachLeakedControls) {
     pControlList->clear();
+    MMutexLocker locker(&s_qCOHashMutex);
+    pControlList->reserve(s_qCOHash.size());
     for (auto it = s_qCOHash.constBegin(); it != s_qCOHash.constEnd(); ++it) {
-        QSharedPointer<ControlDoublePrivate> pControl = it.value();
-        if (!pControl.isNull()) {
+        auto pControl = it.value().lock();
+        if (pControl) {
             pControlList->push_back(pControl);
         }
     }
-    s_qCOHashMutex.unlock();
+    if (detachLeakedControls) {
```

Comment:
"detachLeakedControls" is hard to understand because leaked is only valid in the caller context.  Can't we just call it "clear" or "take" ?